### PR TITLE
Wrap writing to files into function to report any errors 

### PR DESCRIPTION
**Pull request by BevapDin**
Two wrapper classes (containing a `std::ofstream`) do error checking upon opening the file and upon closing it. They throw exceptions (`std::runtime_error`), which are caught together with the errors from the JSON classes (if any) by catching `std::exception`. The check upon closing the stream ensures that all data has been formatted and written properly.

Two functions encapsulate the whole process even further: they take the path of the file and a `std::function<void(std::ostream&)` (called with the file stream, does the actual output).

One function/class is for exclusive opening (see mapshiaring.cpp) instead of normal opening.

### CODE REVIEW

File: src/worldfactory.cpp [hunk 1]
```diff
@@ -257,45 +258,37 @@ bool worldfactory::save_world(WORLDPTR world, bool is_conversion)
...
+            JsonOut jout( fout );
 
-        jout.start_array();
+            jout.start_array();
 
-        for( auto &elem : world->WORLD_OPTIONS ) {
-            if( elem.second.getDefaultText() != "" ) {
-                jout.start_object();
+            for( auto &elem : world->WORLD_OPTIONS ) {
+                if( elem.second.getDefaultText() != "" ) {
```
Comment (codemime) [inline src/worldfactory.cpp]: Why not `empty()`?

Comment (BevapDin) [inline src/worldfactory.cpp]: Preexisting code. Doesn't even show up in the diff without white spaces: https://github.com/CleverRaven/Cataclysm-DDA/pull/16426/files?w=1

File: src/cata_utility.cpp [hunk 2]
```diff
@@ -315,3 +317,79 @@ float multi_lerp( const std::vector<std::pair<float, float>> &points, float x )
...
+{
+    try {
+        ofstream_wrapper fout( path );
+        writer( fout.stream() );
+        fout.close();
+        return true;
+
+    } catch( const std::exception &err ) {
+        if( fail_message ) {
+            popup( _( "Failed to write %1$s to \"%2$s\": %3$s" ), fail_message, path.c_str(), err.what() );
```
Comment (illi-kun) [inline src/cata_utility.cpp]: Why this message is translatable? What is the difference between it & other debug messages?

Comment (BevapDin) [inline src/cata_utility.cpp]: The user may encounter this message (file system is full, or insufficient permissions to write files), they indicate a problem on the their system, not necessarily in the program. They need to be useful for the user.

Debug messages are not supposed to appear at all and indicate a problem in the program, they need to be useful for the programmer (and the user is supposed to forward them verbatim).

I made the exception messages translatable as well.

Comment (illi-kun) [inline src/cata_utility.cpp]: It makes sense, thanks for explanation.

**Comment by Coolthulhu** [comment]
Debugmsg:
"Overmap (0,0) failed to load: line 2,12: expected string but got '['"
Happens every time I save and load a world (old saves don't produce the debugmsg).

**Comment by BevapDin** [comment]
Fixed and also removed the accidentally included file.

**Pull request merged**